src/ui/win32/DialogBase.cpp

```diff
@@ -0,0 +1,128 @@
+#include "DialogBase.hh"
+
+#include "RA_Core.h" // g_RAMainWnd, g_hThisDLLInst
+
+namespace ra {
+namespace ui {
+namespace win32 {
+
+DialogBase::DialogBase(ra::ui::WindowViewModelBase& vmWindow) noexcept
+    : m_vmWindow(vmWindow), m_bindWindow(vmWindow)
+{
+}
+
+DialogBase::~DialogBase()
+{
+    if (m_hWnd)
+    {
+        DestroyWindow(m_hWnd);
+        m_hWnd = nullptr;
+    }
+}
+
+static INT_PTR CALLBACK StaticDialogProc(HWND hDlg, UINT uMsg, WPARAM wParam, LPARAM lParam)
+{
+    DialogBase* pDialog = reinterpret_cast<DialogBase*>(GetWindowLongPtr(hDlg, DWLP_USER));
+    if (pDialog == nullptr)
+        return DefWindowProc(hDlg, uMsg, wParam, lParam);
+
+    INT_PTR result = pDialog->DialogProc(hDlg, uMsg, wParam, lParam);
+
+    if (uMsg == WM_DESTROY)
+        SetWindowLongPtr(hDlg, DWLP_USER, 0L);
+
+    return result;
+}
+
+HWND DialogBase::CreateDialogWindow(LPTSTR sResourceId, IDialogPresenter* pDialogPresenter)
+{
+    m_hWnd = CreateDialog(g_hThisDLLInst, sResourceId, g_RAMainWnd, &StaticDialogProc);
+    if (m_hWnd)
+    {
+        SetWindowLongPtr(m_hWnd, DWLP_USER, reinterpret_cast<LONG>(this));
+        m_pDialogPresenter = pDialogPresenter;
+        m_bindWindow.SetHWND(m_hWnd);
+    }
+
+    return m_hWnd;
+}
+
+INT_PTR CALLBACK DialogBase::DialogProc(_UNUSED HWND hDlg, UINT uMsg, WPARAM wParam, LPARAM lParam)
+{
+    switch (uMsg)
+    {
+        case WM_INITDIALOG:
+            OnInitDialog();
```

Comment:
If you want to get all the values you can get from the kernel there are some message crackers. I'll do em one by one. If you don't want to to them now that's fine. Google "Message Cracker Wizard" if you want an easier time. I was thinking of doing something like this later as well but need to make sure the core is "strong" enough to support the revised UI layer. 

These prototypes will allow you to use them in `HANDLE_MSG`.
`BOOL OnInitDialog(HWND hwnd, HWND hwndFocus, LPARAM lParam);`

Reply:
I'm torn between exposing everything and exposing only what's needed. For simplicity, I'm going to choose the latter. If we find the need for these parameters later, we can change the signature.

`hWnd` for this, and the other functions will be accessible through `m_hWnd`

Reply:
That sounds like a good approach to me.